### PR TITLE
feat: Add Gemini Realtime provider implementing IRealtimeClient/IRealtimeClientSession

### PR DESCRIPTION
**Pull request by tarekgh**
## Summary

Adds a Gemini Live API provider implementing the `Microsoft.Extensions.AI` Realtime abstractions (`IRealtimeClient` / `IRealtimeClientSession`), enabling real-time audio, text, and function-calling conversations with Gemini models through the standardized MEAI interface.

This PR also updates the repository to depend on the **official** `Microsoft.Extensions.AI.Abstractions 10.4.1` NuGet package (replacing the private `10.5.0-dev` builds).

### AOT Compatibility

This PR includes changes to make the Google GenAI SDK fully compatible with AOT (Ahead-of-Time) compilation. This was done to avoid duplicating the SDK's WebSocket and JSON protocol code in the realtime provider, since the MEAI realtime wrapper (`GoogleGenAIRealtimeClient` / `GoogleGenAIRealtimeSession`) delegates all serialization and WebSocket communication to the core SDK's `Live.cs` / `AsyncSession` / `Transformers.cs` / `LiveConverters.cs` code paths.

**AOT changes:**
- Added `GenAIJsonContext` — a source-generated `JsonSerializerContext` with `[JsonSerializable]` entries for 93 root types (nested property types are auto-discovered by the generator)
- Wired the source-gen context into `JsonConfig.JsonSerializerOptions` with `DefaultJsonTypeInfoResolver` fallback for non-AOT scenarios (anonymous types, user-provided types)
- Added compact `InternalSerializerOptions` for intermediate serialize-then-parse round-trips to avoid `WriteIndented` overhead on internal transforms
- Updated all ~90 bare `JsonSerializer.Serialize/Deserialize` call sites across the SDK to use configured options with source-gen metadata
- Added `System.Text.Json` PackageReference for source generator support on both `netstandard2.0` and `net8.0` targets
- Added `AotJsonContextTest.cs` with 4 tests verifying context coverage, nested type auto-discovery, and serialization round-trips

## Usage Example

```csharp
using Google.GenAI;
using Microsoft.Extensions.AI;

// Simple: create with API key and model
IRealtimeClient realtimeClient = new GoogleGenAIRealtimeClient(
    apiKey, "gemini-3.1-flash-live-preview");

// Or advanced: create from an existing Google.GenAI.Client instance
// var genaiClient = new Client(apiKey: apiKey);
// IRealtimeClient realtimeClient = new GoogleGenAIRealtimeClient(genaiClient, "gemini-3.1-flash-live-preview");

// Or use the extension method
// IRealtimeClient realtimeClient = genaiClient.AsIRealtimeClient("gemini-3.1-flash-live-preview");

// Define a tool for function calling
AIFunction getWeather = AIFunctionFactory.Create(
    (string location) =>
        location.ToLowerInvariant() switch
        {
            var l when l.Contains("seattle")       => $"The weather in {location} is rainy, 55F",
            var l when l.Contains("new york")      => $"The weather in {location} is cloudy, 70F",
            var l when l.Contains("san francisco") => $"The weather in {location} is foggy, 60F",
            _                                      => $"Sorry, I don't have weather data for {location}."
        },
    "GetWeather",
    "Gets the current weather for a given location");

// Configure session options
var sessionOptions = new RealtimeSessionOptions
{
    Instructions = "You are a helpful assistant.",
    Voice = "Puck",
    OutputModalities = ["audio"],
    Tools = [getWeather],
    InputAudioFormat = new RealtimeAudioFormat("audio/pcm;rate=16000", 16000),
    TranscriptionOptions = new TranscriptionOptions(),
    VoiceActivityDetection = new VoiceActivityDetectionOptions
    {
        Enabled = true,
        AllowInterruption = true,
    },
};

// Optionally wrap with function invocation middleware
var clientBuilder = new RealtimeClientBuilder(realtimeClient)
    .UseFunctionInvocation();
IRealtimeClient wrappedClient = clientBuilder.Build();

// Create a session and start streaming
await using var session = await wrappedClient.CreateSessionAsync(sessionOptions);

// Start listening for server messages in the background
_ = Task.Run(async () =>
{
    await foreach (var message in session.GetStreamingResponseAsync(cancellationToken))
    {
        switch (message)
        {
            case OutputTextAudioRealtimeServerMessage audio
                when audio.Type == RealtimeServerMessageType.OutputAudioDelta:
                // Play audio - audio.Audio contains base64-encoded PCM audio
                PlayAudio(Convert.FromBase64String(audio.Audio));
                break;

            case OutputTextAudioRealtimeServerMessage text
                when text.Type == RealtimeServerMessageType.OutputAudioTranscriptionDelta:
                Console.Write(text.Text);
                break;

            case InputAudioTranscriptionRealtimeServerMessage transcription:
                Console.WriteLine($"You said: {transcription.Transcription}");
                break;

            case ResponseCreatedRealtimeServerMessage response:
                if (response.Usage != null)
                    Console.WriteLine($"Tokens - In: {response.Usage.InputTokenCount}, Out: {response.Usage.OutputTokenCount}");
                break;

            case ResponseOutputItemRealtimeServerMessage item:
                if (item.Item is RealtimeConversationItem convItem)
                    foreach (var content in convItem.Contents)
                        if (content is FunctionCallContent fc)
                            Console.WriteLine($"Function call: {fc.Name}({string.Join(", ", fc.Arguments?.Select(a => $"{a.Key}={a.Value}") ?? [])})");
                break;

            case ErrorRealtimeServerMessage error:
                Console.Error.WriteLine($"Error: {error.Error?.Message}");
                break;
        }
    }
});

// Send audio from microphone (e.g., 16kHz PCM)
var audioContent = new DataContent($"data:audio/pcm;base64,{Convert.ToBase64String(audioBytes)}");
await session.SendAsync(new InputAudioBufferAppendRealtimeClientMessage(audioContent));
await session.SendAsync(new InputAudioBufferCommitRealtimeClientMessage());
await session.SendAsync(new CreateResponseRealtimeClientMessage());
```

### New Files
- **`GoogleGenAIRealtimeClient.cs`** — `IRealtimeClient` implementation that wraps a `Google.GenAI.Client` and creates realtime sessions via the Gemini Live API. Includes a convenience constructor accepting just an API key and model ID.
- **`GoogleGenAIRealtimeSession.cs`** — `IRealtimeClientSession` implementation that manages the WebSocket connection, audio buffering, message mapping, and function call orchestration.
- **`GoogleGenAIRealtimeTest.cs`** — 118 unit tests covering the full surface area.
- **`GenAIJsonContext.cs`** — Source-generated JSON serialization context for AOT compatibility.
- **`AotJsonContextTest.cs`** — 4 unit tests verifying AOT source-gen coverage and round-trip correctness.

### Modified Files
- **`GoogleGenAIExtensions.cs`** — Added `AsIRealtimeClient()` extension method.
- **`Directory.Packages.props`** — Updated `Microsoft.Extensions.AI.Abstractions` from `10.5.0-dev` → `10.4.1`.
- **`Google.GenAI.csproj`** — Added `System.Text.Json` PackageReference for source generator support.
- **`JsonConfig.cs`** — Dual options: `JsonSerializerOptions` (indented, for API output) + `InternalSerializerOptions` (compact, for internal transforms). Both use source-gen context with reflection fallback.
- **`Live.cs`** — Minor adjustment to expose `AsyncSession` for the realtime provider; added serialization options.
- **`Batches.cs, Caches.cs, Files.cs, Models.cs, Operations.cs, Tunings.cs`** — Updated all bare `JsonSerializer` calls to use configured options (intermediate → `InternalSerializerOptions`, HTTP body/response → `JsonSerializerOptions`).
- **`Transformers.cs, Common.cs, TokensConverters.cs`** — Updated bare `JsonSerializer` calls to use configured options.
- All `packages.lock.json` files regenerated.

## Features

- ✅ **Audio streaming** — Append/commit pattern with automatic frame splitting (32KB max), ActivityStart/ActivityEnd framing
- ✅ **Voice Activity Detection (VAD)** — Configurable server-side VAD or manual client-controlled boundaries
- ✅ **Text conversations** — Send text messages and receive text/audio responses
- ✅ **Function calling** — Full tool invocation support with the `FunctionInvokingRealtimeSession` middleware; tool responses are batched into a single `SendToolResponseAsync` call
- ✅ **Transcription** — Input and output audio transcription
- ✅ **Thread-safe sends** — `SemaphoreSlim` serializes all WebSocket sends, safe for concurrent middleware + caller usage
- ✅ **Graceful disposal** — Race-safe dispose with proper exception handling for in-flight operations
- ✅ **SetupComplete handshake** — `CreateSessionAsync` waits for the server's `SetupComplete` acknowledgment before returning, ensuring tools and modalities are fully configured before the caller sends audio or text
- ✅ **Convenience constructor** — `GoogleGenAIRealtimeClient(string apiKey, string? defaultModelId)` for simple setup without manually creating a `Client`
- ✅ **AOT compatible** — Source-generated JSON serialization for all SDK types, verified by dedicated tests

## Key Design Decisions

1. **SetupComplete handshake** — The Google SDK's `ConnectAsync` sends the setup config but returns immediately without waiting for the server's `SetupComplete` acknowledgment. Our `CreateSessionAsync` drains this message before returning, ensuring the session is fully ready (tools configured, modalities set). Without this, function calling fails when the user speaks immediately after connecting.

2. **Tool response batching** — The MEAI `FunctionInvokingRealtimeSession` middleware sends separate `CreateConversationItem` per function result. Gemini expects all results in one `SendToolResponseAsync` call. The provider buffers results and flushes them as a single batch when `CreateResponse` arrives.

3. **TurnComplete suppression after tool responses** — After `SendToolResponseAsync`, Gemini automatically continues generating. Sending `client_content` with `turn_complete: true` causes the server to close the WebSocket. The provider tracks this via `_lastSendWasToolResponse` and skips TurnComplete accordingly.

4. **Null function call ID guard** — If a function call arrives without an ID, a synthetic GUID is generated to ensure the call-ID-to-function-name mapping always works for the round-trip.

5. **VAD handling** — When VAD is disabled (default), the provider wraps audio commits with explicit `ActivityStart`/`ActivityEnd` framing. When enabled, the server handles speech boundary detection automatically.

6. **Audio buffer cap** — Audio appends are capped at 10 MB to prevent unbounded memory growth. Frames exceeding 32 KB are automatically split.

7. **AOT via source-gen in core SDK** — Rather than reimplementing the WebSocket+JSON protocol in the MEAI wrapper (which would duplicate ~1000 lines of code), we made the core SDK AOT-compatible by adding a `JsonSerializerContext` with source-generated metadata for all types. This ensures the MEAI realtime wrapper can delegate to `Live.cs`/`AsyncSession` without any reflection-based JSON calls on the hot path.

## Test Coverage

122 unit tests covering:
- Client and session lifecycle (construction, disposal, idempotent dispose)
- All message types (audio, text, function calls, transcription, errors)
- Edge cases (null args, empty buffers, concurrent dispose, exception swallowing)
- Function call flow (single/multiple results, batching, flag reset after tool cycle)
- VAD modes (enabled, disabled, default)
- `BuildLiveConnectConfig` mapping (all option combinations)
- AOT source-gen context coverage, nested type discovery, and round-trip correctness

### CODE REVIEW

File: Google.GenAI/Batches.cs
Comment (yyyu-google) [inline Google.GenAI/Batches.cs:17]: this file is not meant to be modified manually.

ditto for all other files with the same code comment.

Comment (tarekgh) [reply Google.GenAI/Batches.cs:17]: @yyyu-google I’ve reverted the changes in these files, which means the SDK is no longer AOT-compatible. We’ll likely need to update the code generators to produce code similar to what was reverted. I’ll leave that decision to you.

**Comment by tarekgh** [comment]
@yyyu-google will you have a chance to complete the review of this PR?

**Comment by yyyu-google** [comment]
Hi @matthew29tang will be reviewing this PR.

File: Google.GenAI/Batches.cs
Comment (matthew29tang) [reply Google.GenAI/Batches.cs:17]: Hi, trying to understand this thread. With the changes reverted, will this PR still build and pass all the tests? Can you describe the changes you need to the autogenerated files?

Comment (tarekgh) [reply Google.GenAI/Batches.cs:17]: > Will this PR still build and pass all tests?

Yes, it should. The revert commit https://github.com/googleapis/dotnet-genai/pull/256/commits/c1d6f5eaa44f38a91dcbf9732351d1e19a33e153 simply removed the `JsonConfig.InternalSerializerOptions` / `JsonConfig.JsonSerializerOptions` parameters from all `JsonSerializer.Serialize()` calls in the 6 auto-generated files, restoring them to bare `JsonSerializer.Serialize(obj)`. This is functionally correct, it just falls back to default reflection-based serialization instead of using the source-generated AOT context.

> What changes are needed to the autogenerated files?

The pattern is straightforward and mechanical. In each auto-generated file, every `JsonSerializer.Serialize(...)` call needs to pass the appropriate options:
  - Internal round-trips (serialize → parse to `JsonNode` for request building): `JsonSerializer.Serialize(parameter, JsonConfig.InternalSerializerOptions)`
  - HTTP body serialization (sent to the API): `JsonSerializer.Serialize(body, JsonConfig.JsonSerializerOptions)`
 
The code generators that produce these files would need to emit the options parameter.

Without these changes, the SDK works but is not AOT-compatible; all JSON serialization in auto-generated code falls back to runtime reflection, which fails in trimmed/AOT-published apps.

Comment (matthew29tang) [reply Google.GenAI/Batches.cs:17]: Thanks for the explanation. After we submit this PR, I can make a follow-up internal change in our code generators that covers the auto-generated files, and you can verify that it has the correct results.

Will some tests in your PR need to be skipped until the changes to the auto-generated files land?

Also, if we add more types to the SDK in the future, is there any manual code changes/tests that need to be done to ensure that they are also AOT compliant? Assuming that the auto-generated files related to the types are correct (applied to all types).

**Comment by matthew29tang** [comment]
I'm trying to look at the diffs and noticed that there are large deletion/additions for the same file but the contents look similar, making it hard to follow the diff. Ex common.cs or live.cs. 

Is there a reason for this (is it linting differences)?

**Comment by tarekgh** [comment]
> I'm trying to look at the diffs and noticed that there are large deletion/additions for the same file but the contents look similar, making it hard to follow the diff. Ex common.cs or live.cs.

Could it be just whitespaces diff? Could you turn on ignoring whitespace in the diff?

File: Google.GenAI/Batches.cs
Comment (tarekgh) [reply Google.GenAI/Batches.cs:17]: > Will some tests in your PR need to be skipped until the changes to the auto-generated files land?

Would any test failures show in the CI results? I can look ensuring which tests need to change or should be disabled? 

> Also, if we add more types to the SDK in the future, is there any manual code changes/tests that need to be done to ensure that they are also AOT compliant? Assuming that the auto-generated files related to the types are correct (applied to all types).

The best action here is to add AOT tests to this repo to ensure catching any issue before merging it.

Comment (matthew29tang) [reply Google.GenAI/Batches.cs:17]: > Would any test failures show in the CI results? I can look ensuring which tests need to change or should be disabled?

So the Github check `[.NET Tests PR Check / unit-e2e-tests (pull_request)](https://github.com/googleapis/dotnet-genai/actions/runs/24540717876/job/74902194247?pr=256)` should run the tests, which does have a green check mark. My concern is rather that the tests that are checking AOT compliance are passing despite us being no longer AOT compliant (from your reverted commit). If they are meant to pass with the current state of the PR, then what exactly is Google.GenAI.Tests/AotJsonContextTest.cs testing?

>  The best action here is to add AOT tests to this repo to ensure catching any issue before merging it.

To clarify, I am asking about the maintenance effort moving forward. The SDK will continue to add more types over time. Are there additional steps involved every time to ensure each new class added is AOT compliant, assuming the generator handles the auto-generated files correctly?

**Comment by matthew29tang** [comment]
> Could it be just whitespaces diff? Could you turn on ignoring whitespace in the diff?

Thanks, ignoring whitespace diff helps me review the PR. But it would still be good to figure out why the whitespace diff is being introduced before merging the PR. Are you using a different linter? Or are you suggesting that the existing files not linter compliant right now?

**Comment by tarekgh** [comment]
> it would still be good to figure out why the whitespace diff 

It was the classical issue of having CRLF against LF. I cleaned up these and you shouldn't see the whitespace diffs.

File: Google.GenAI/Batches.cs
Comment (tarekgh) [reply Google.GenAI/Batches.cs:17]: > My concern is rather that the tests that are checking AOT compliance are passing despite us being no longer AOT compliant (from your reverted commit). If they are meant to pass with the current state of the PR, then what exactly is Google.GenAI.Tests/AotJsonContextTest.cs testing?

The tests still pass because they test a **partial** AOT story. They verify that the `GenAIJsonContext` source-generated context has metadata for all types and can round-trip serialize/deserialize without reflection. That part is valid and working.

The gap is that the 7 reverted auto-generated files (`Batches.cs`, `Caches.cs`, `Files.cs`, `Models.cs`, `Operations.cs`, `TokensConverters.cs`, `Tunings.cs`) now use bare `JsonSerializer.Serialize(obj)` calls without passing `JsonConfig.InternalSerializerOptions`/`JsonConfig.JsonSerializerOptions`. Those bare calls use the global default options which don't include the source-gen context, so in a true AOT deployment, those call sites would fail.

The existing tests don't catch this because they test the context directly (`GenAIJsonContext.Default.GetTypeInfo(...)`) rather than exercising the actual serialization paths in the auto-generated files. A true AOT validation test (e.g., publishing as native AOT and running) would be needed for full end-to-end coverage.

> To clarify, I am asking about the maintenance effort moving forward. The SDK will continue to add more types over time. Are there additional steps involved every time to ensure each new class added is AOT compliant, assuming the generator handles the auto-generated files correctly?

Yes, two things are needed when new types are added:

1. **`GenAIJsonContext.cs`** - Any new root type that gets serialized/deserialized needs a `[JsonSerializable(typeof(NewType))]` attribute added. Nested types (properties of registered types) are auto-discovered by the source generator, so only top-level entry points need to be registered. The existing `AotJsonContextTest` would catch any missing types here.

2. **Auto-generated files** - Every `JsonSerializer.Serialize()`/`Deserialize()` call must pass `JsonConfig.InternalSerializerOptions` or `JsonConfig.JsonSerializerOptions` so the source-gen context is actually used at those call sites. This is the part that needs the code generator update on your side.

The maintenance effort for (1) is light - it's just adding one attribute line per new root type, and the test catches omissions. For (2), once the code generator is updated to emit the options parameter, it should be automatic for all future types with no manual steps needed.

**Regarding test updates after the code generator fix lands:**

Once the auto-generated files are updated to pass the serializer options, the existing `AotJsonContextTest` tests will continue to pass as-is (they validate context coverage which is already correct). However, to prevent future regressions, here is what I'd recommend adding:

1. **A source scan test** - verify that no bare `JsonSerializer.Serialize(obj)` or `JsonSerializer.Deserialize<T>(json)` calls exist without passing options. This could be a simple regex scan of all `.cs` files ensuring every `JsonSerializer.Serialize`/`Deserialize` call includes a `JsonConfig.*Options` or `GenAIJsonContext.Default.*` parameter. This would catch regressions if someone adds new code without the options.

2. **A native AOT publish test** (ideal but heavier) - a CI step that publishes a small app with `dotnet publish -r <rid> /p:PublishAot=true` and runs it. This is the only way to truly catch AOT failures since reflection-based serialization works fine in normal testing but fails at AOT trim time.

3. **`GenAIJsonContext.cs` updates** - if the generator introduces new types, corresponding `[JsonSerializable(typeof(NewType))]` entries need to be added. The existing `GenAIJsonContext_CoversAllDirectlySerializedTypes` test would catch omissions, but its `requiredTypes` array would also need updating to include the new types.

Option (1) is lightweight and high-value. Option (2) is the gold standard but requires CI infrastructure changes. I can help with either after the generator update lands.

File: Google.GenAI/GoogleGenAIRealtimeClient.cs [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+  {
+    _client = client ?? throw new ArgumentNullException(nameof(client));
+    _defaultModelId = defaultModelId;
+  }
+
+  /// <summary>Initializes a new instance using an API key.</summary>
+  /// <param name="apiKey">The Google GenAI API key.</param>
+  /// <param name="defaultModelId">The default model to use for realtime sessions.</param>
+  /// <exception cref="ArgumentNullException"><paramref name="apiKey"/> is <see langword="null"/> or empty.</exception>
+  public GoogleGenAIRealtimeClient(string apiKey, string? defaultModelId = null)
```
Comment (matthew29tang) [inline Google.GenAI/GoogleGenAIRealtimeClient.cs:48]: I noticed that we have a constructor here for apikey but not one for project/location. Should we also have one for project/location?

File: Google.GenAI/Batches.cs
Comment (matthew29tang) [reply Google.GenAI/Batches.cs:17]: Sounds good, we can do Option 1 for the source scan test after the generator /auto-generated file changes land

File: Google.GenAI/GoogleGenAIRealtimeClient.cs (hunk 1 above)
Comment (tarekgh) [reply Google.GenAI/GoogleGenAIRealtimeClient.cs:48]: I added the constructor. Thanks!

**Comment by matthew29tang** [comment]
Can you resolve the merge conflicts? We recently bumped the version of MEAI extensions in https://github.com/googleapis/dotnet-genai/commit/81236bc9f04d776dcb13a88d9ec3fca392edd557

File: Google.GenAI/GoogleGenAIRealtimeClient.cs (hunk 1 above)
Comment (matthew29tang) [reply Google.GenAI/GoogleGenAIRealtimeClient.cs:48]: Could you also add a test for this project/location constructor?

File: local-packages/Google.GenAI.1.0.0-local.nupkg
Comment (matthew29tang) [inline local-packages/Google.GenAI.1.0.0-local.nupkg:1]: What's the purpose of this file? Saw it is a new file being added in this PR.

File: nuget.config
Comment (matthew29tang) [inline nuget.config:1]: What's the purpose of this file? Saw it is a new file being added in this PR.

File: local-packages/Google.GenAI.1.0.0-local.nupkg
Comment (tarekgh) [reply local-packages/Google.GenAI.1.0.0-local.nupkg:1]: This was not intended to merge; I was building for local use. I'll clean this up.

File: nuget.config
Comment (tarekgh) [reply nuget.config:1]: ditto

File: local-packages/Google.GenAI.1.0.0-local.nupkg
Comment (tarekgh) [reply local-packages/Google.GenAI.1.0.0-local.nupkg:1]: Done

File: Google.GenAI/GoogleGenAIRealtimeClient.cs (hunk 1 above)
Comment (tarekgh) [reply Google.GenAI/GoogleGenAIRealtimeClient.cs:48]: done

**Review by matthew29tang — APPROVED** [review]
Thanks for the prompt turnarounds! Just going to check on a few more things with my team before merging this in, but otherwise it looks good!

**Review by matthew29tang — APPROVED** [review]

**Comment by matthew29tang** [comment]
Merged as https://github.com/googleapis/dotnet-genai/commit/2b92e8588548a3fa578d5206b77dbd3717ee6296, thanks!

**Comment by tarekgh** [comment]
Thanks, @matthew29tang, for your continued follow-up on this. Do we have an estimated timeline for when this will be released to NuGet?

**Comment by matthew29tang** [comment]
Likely Wed June 10. I'll make a follow-up shortly with the generator changes (and the auto-generated file diff) to cover what was removed in https://github.com/googleapis/dotnet-genai/commit/c1d6f5eaa44f38a91dcbf9732351d1e19a33e153.
